exch: fix compile error if renameat2 is not present

ThomasDevoogdt opened the pull request:
https://www.man7.org/linux/man-pages/man2/rename.2.html Available since Linux 3.15 or glibc 2.28,
add a fallback for older systems.

masatake commented:
I think `exch` should fail if the runtime platform doesn't provide `RENAME_EXCHANGE`; I assume `exch` provides atomic exchange.

ThomasDevoogdt commented:
> I think `exch` should fail if the runtime platform doesn't provide `RENAME_EXCHANGE`; I assume `exch` provides atomic exchange.

Understood, I just dropped that part, and added a -1 return.

masatake commented on misc-utils/exch.c:
```diff
@@ -26,14 +26,22 @@
+	rc = syscall (SYS_renameat2,
+		      dirfd, oldpath, dirfd, newpath, RENAME_EXCHANGE);
+#else
+	rc = -1;
```
What do you think setting `ENOSYS` to `errno` in addition to setting -1 to `rc`.
So the next commit that you proposed in this pull request can be removed.

I'm not sure what I wrote here is a good way.
@karelzak @t-8ch may give us hints.

ThomasDevoogdt replied:
Ah, that success came from errno.

errno = ENOSYS; gives me this:
```
exch: failed to exchange "a" and "b": Function not implemented
```

which is indeed better.

masatake commented on misc-utils/exch.c:
```diff
@@ -26,14 +26,23 @@
-{
-	return syscall (SYS_renameat2, olddirfd, oldpath, newdirfd, newpath, flags);
-}
+static inline int rename_exchange(int dirfd,
```
I guess there can be a build environment where SYS_renameat2 is defined but RENAME_EXCHANGE is not.

How about adding the definition for RENAME_EXCHANGE?

```
#ifndef RENAME_EXCHANGE
#define RENAME_EXCHANGE		(1 << 1)	/* Taken from linux/include/uapi/linux/fs.h */
#endif
```

We can do the same for SYS_renameat2. However, I think it is over-killing.

Let's see how the maintainer says :-)

ThomasDevoogdt commented:
RENAME_EXCHANGE is already defined a few lines above the method code block.

Op vr 7 mrt 2025 om 20:22 schreef Masatake YAMATO ***@***.***
>:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In misc-utils/exch.c
> <https://github.com/util-linux/util-linux/pull/3448#discussion_r1985574820>
> :
>
> > @@ -26,14 +26,23 @@
>  # define RENAME_EXCHANGE (1 << 1)
>  #endif
>
> -#if !defined(HAVE_RENAMEAT2) && defined(SYS_renameat2)
> -static inline int renameat2(int olddirfd, const char *oldpath,
> -			    int newdirfd, const char *newpath, unsigned int flags)
> -{
> -	return syscall (SYS_renameat2, olddirfd, oldpath, newdirfd, newpath, flags);
> -}
> +static inline int rename_exchange(int dirfd,
>
> I guess there can be a build environment where SYS_renameat2 is defined
> but RENAME_EXCHANGE is not.
>
> How about adding the definition for RENAME_EXCHANGE?
>
> #ifndef RENAME_EXCHANGE
> #define RENAME_EXCHANGE		(1 << 1)	/* Taken from linux/include/uapi/linux/fs.h */
> #endif
>
> We can do the same for SYS_renameat2. However, I think it is over-killing.
>
> Let's see how the maintainer says :-)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/util-linux/util-linux/pull/3448#pullrequestreview-2668130236>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACDAFBI75GGUU7DWZXWPIML2THWVXAVCNFSM6AAAAABYOV7IMOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMNRYGEZTAMRTGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

masatake commented:
> RENAME_EXCHANGE is already defined a few lines above the method code block.

```
cd094a0588 (Masatake YAMATO 2023-12-07 25) #ifndef RENAME_EXCHANGE
```

sigh. Aging is in progress.

t-8ch reviewed (approved):
On old kernels this produces an `exch` binary that is completely useless which is unfortunate.
OTOH those are already out of the proper support window anyways, so it shouldn't matter.

karelzak commented:
> On old kernels this produces an `exch` binary that is completely useless which is unfortunate. 

In the ./configure, a dependence on SYS_renameat2 is defined.

t-8ch commented:
> In the ./configure, a dependence on SYS_renameat2 is defined.

Then the -1/ENOSYS fallback is unnecessary.

karelzak commented:
I also did small changes to the code https://github.com/util-linux/util-linux/commit/6bef160562189a72c6f5c779cb7e06e72a0fe6bc

karelzak commented:
> > In the ./configure, a dependence on SYS_renameat2 is defined.
> 
> Then the -1/ENOSYS fallback is unnecessary.

Let's keep the code robust (and readable) independently on the rest :-)
